tikv-client/src/main/java/com/pingcap/tikv/operation/KVErrorHandler.java

```diff
@@ -145,30 +159,42 @@ public boolean handleResponseError(BackOffer backOffer, RespT resp) {
 
         return true;
       } else if (error.hasStoreNotMatch()) {
-        logger.warn(String.format("Store Not Match happened with region id %d, store id %d",
-            ctxRegion.getId(),
-            ctxRegion.getLeader().getStoreId()));
-
-        invalidateRegionStoreCache(ctxRegion);
-        recv.onStoreNotMatch();
+        // this error is reported from raftstore:
+        // store_id requested at the moment is inconsistent with that expected
+        // Solution：re-fetch from PD
+        long storeId = ctxRegion.getLeader().getStoreId();
+        logger.warn(String.format("Store Not Match happened with region id %d, store id %d", ctxRegion.getId(), storeId));
+
+        this.regionManager.invalidateStore(storeId);
```

Comment:
Just invalidate store cache is enough for StoreNotMatch error. See [this line in tidb](https://github.com/pingcap/tidb/blob/a8012590dcce643db7a837266f6849b9435ad57e/store/tikv/region_request.go#L207)